pkg/client/client_test.go

```diff
@@ -19,152 +19,942 @@ import (
 	"reflect"
 	"testing"
 
+	secv1 "github.com/openshift/api/security/v1"
 	"github.com/openshift/cluster-monitoring-operator/pkg/manifests"
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
 	rbacv1 "k8s.io/api/rbac/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/client-go/kubernetes"
+
+	monv1 "github.com/coreos/prometheus-operator/pkg/apis/monitoring/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
 	"k8s.io/client-go/kubernetes/fake"
+
+	monfake "github.com/coreos/prometheus-operator/pkg/client/versioned/fake"
+	ossfake "github.com/openshift/client-go/security/clientset/versioned/fake"
+)
+
+const (
+	ns    = "openshift-monitoring"
+	nsUWM = "openshift-user-workload-monitoring"
 )
 
+func TestMergeMetadata(t *testing.T) {
+	testCases := []struct {
+		name     string
+		expected map[string]string
+		new      map[string]string
+		old      map[string]string
+	}{
+		{
+			name: "new annotation and label addition",
+			expected: map[string]string{
+				"old": "value",
+				"new": "value",
+			},
+			old: map[string]string{
+				"old": "value",
+			},
+			new: map[string]string{
+				"new": "value",
+			},
+		},
+		{
+			name: "immutable annotation and label values",
+			expected: map[string]string{
+				"key": "old",
+			},
+			old: map[string]string{
+				"key": "old",
+			},
+			new: map[string]string{
+				"key": "new",
+			},
+		},
+		{
+			name: "annotation and label removal",
+			expected: map[string]string{
+				"key": "value",
+			},
+			old: map[string]string{
+				"key": "value",
+			},
+			new: map[string]string{
+				"monitoring.openshift.io/new": "value",
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(st *testing.T) {
+			newMeta := metav1.ObjectMeta{
+				Annotations: tc.new,
+				Labels:      tc.new,
+			}
+			oldMeta := metav1.ObjectMeta{
+				Annotations: tc.old,
+				Labels:      tc.old,
+			}
+
+			mergeMetadata(&oldMeta, newMeta)
+
+			if !reflect.DeepEqual(oldMeta.Annotations, tc.expected) || !reflect.DeepEqual(oldMeta.Labels, tc.expected) {
+				t.Logf("test for %s failed", tc.name)
+				t.Fail()
```

Comment:
nits:
1. asserting Annotations and Labels should be separate
2. we don't need to hard-fail, `t.Errorf` is good enough and allows proceeding, because this is a unit test, not an integration test.
3. let's use idiomatic test error messages, i.e. `t.Errorf("expected old annotations %q, got %q", tc.expected, oldMeta.Annotations)`

Reply:
I separated asserting annotations and labels and used more idiomatic msgs. As for `t.Errorf()` it is functionally the same as `t.Logf()` followed by `t.Fail()` as mentioned in golang docs https://golang.org/pkg/testing/#T.Errorf and I changed it to `t.Errorf` for cleaner code.